### PR TITLE
DBI-76: Upgrading Jena libraries to 4.9.0

### PR DESCRIPTION
**Pull request by abika5**

### CODE REVIEW

File: src/main/java/com/marklogic/mapreduce/test/CustomQuery.java [hunk 1]
```diff
@@ -17,7 +17,7 @@
 
 import java.io.IOException;
 
-import javax.xml.soap.Node;
+import jakarta.xml.soap.Node;
```
Comment (rjrudin) [inline src/main/java/com/marklogic/mapreduce/test/CustomQuery.java:20]: Seems like this class should be moved to src/test/java if it's intended to be a test?

**Review by rjrudin — COMMENTED** [review]
Just noting that I'm glad to see this, just noticed the log4j vulnerability in a Snyk report.

File: pom.xml [hunk 2]
```diff
@@ -100,6 +101,7 @@
             <classpathDependencyExclude>ch.qos.logback:logback-classic</classpathDependencyExclude>
             <classpathDependencyExclude>org.slf4j:slf4j-reload4j</classpathDependencyExclude>
             <classpathDependencyExclude>ch.qos.reload4j:reload4j</classpathDependencyExclude>
+            <classpathDependencyExclude>xml-apis:xml-apis</classpathDependencyExclude>
```
Comment (yunzvanessa) [inline pom.xml:104]: What is the purpose here to exclude xml-apis here?

Comment (abika5) [reply pom.xml:104]: It was a redundant addition for us as we moved to JDK 11. JDK 11 has the xml-apis included in it. The Xerces jar has this as the compile dependency and had an issue for multiple sources. So I excluded this.

File: src/main/java/com/marklogic/contentpump/RDFReader.java [hunk 3]
```diff
@@ -1075,29 +1046,25 @@ public boolean failed() {
...
-            }
+        public void run() {
             try {
-                ErrorHandler handler = new ParserErrorHandler(fsname);
+                ErrorHandler handler =  ErrorHandlerFactory.errorHandlerStd(ErrorHandlerFactory.stdLogger);
                 ParserProfile prof = RiotLib.profile(lang, fsname, handler);
-                parser.setProfile(prof);
-                parser.parse();
+                if (lang == Lang.TRIG) {
+                    rdfInputStream = AsyncParser.of(in, lang, fsname).streamQuads();
```
Comment (yunzvanessa) [inline src/main/java/com/marklogic/contentpump/RDFReader.java:1055]: Could you please explain why we are using different apis for TRIG and NQUADs?

File: src/main/java/com/marklogic/contentpump/RDFReader.java [hunk 4]
```diff
@@ -1075,29 +1046,25 @@ public boolean failed() {
...
+                    rdfIter = rdfInputStream.iterator();
+                } else if (lang == Lang.NTRIPLES) {
```
Comment (yunzvanessa) [inline src/main/java/com/marklogic/contentpump/RDFReader.java:1057]: Also ntriples and the else case?

File: src/main/java/com/marklogic/contentpump/RDFReader.java [hunk 5]
```diff
@@ -670,8 +639,15 @@ public boolean nextInMemoryKeyValue() throws IOException, InterruptedException {
 
     public boolean nextInMemoryTripleKeyValue() throws IOException, InterruptedException {
         if(statementIter == null) return false;
-        if (!statementIter.hasNext()) {
-            hasNext = false;
+        try {
```
Comment (yunzvanessa) [inline src/main/java/com/marklogic/contentpump/RDFReader.java:642]: What is the purpose of adding a try-catch here?

File: src/main/java/com/marklogic/contentpump/RDFReader.java [hunk 6]
```diff
@@ -708,16 +684,23 @@ public boolean nextInMemoryQuadKeyValue() throws IOException, InterruptedExcepti
...
         if(statementIter == null) return false;
-        while (!statementIter.hasNext()) {
-            if (graphNameIter.hasNext()) {
-                collection = graphNameIter.next();
-                statementIter = dataset.getNamedModel(collection).listStatements();
-            } else {
-                hasNext = false;
-                collection = null;
-                return false;
+        try {
```
Comment (yunzvanessa) [inline src/main/java/com/marklogic/contentpump/RDFReader.java:687]: Same question here

**Review by yunzvanessa — CHANGES_REQUESTED** [review]

File: pom.xml (hunk 2 above)
Comment (abika5) [reply pom.xml:104]: xml-apis is not in the pom file as a dependency but has an entry in bindist files for mlcp and mapr. I can remove it from the bindist files

File: pom.xml [hunk 7]
```diff
@@ -740,7 +798,7 @@
     <dependency>
       <groupId>com.fasterxml.woodstox</groupId>
       <artifactId>woodstox-core</artifactId>
-      <version>6.4.0</version>
+      <version>6.5.1</version>
```
[the remark this answers is not in the document]
Comment (abika5) [reply pom.xml:801]: yes, I will add it

File: src/assemble/bindist.xml [hunk 8]
```diff
@@ -116,6 +116,11 @@
         <include>org.apache.jena:jena-arq:jar:*</include>
         <include>org.apache.jena:jena-core:jar:*</include>
         <include>org.apache.jena:jena-iri:jar:*</include>
+        <include>org.apache.jena:jena-base:jar:*</include>
+        <include>org.apache.jena:jena-shaded-guava:jar:*</include>
+        <include>org.apache.thrift:libthrift:jar:*</include>
```
[the remark this answers is not in the document]
Comment (abika5) [reply src/assemble/bindist.xml:121]: Libthrift library is used on runtime during asyncparser calls. So it has to be added to the lib folder for the commandline execution.

File: src/main/java/com/marklogic/contentpump/CompressedRDFReader.java [hunk 9]
```diff
@@ -59,7 +50,6 @@ public class CompressedRDFReader<VALUEIN> extends RDFReader<VALUEIN> {
     private InputStream zipIn;
     private ZipEntry currZipEntry;
     private CompressionCodec codec;
-    private ExecutorService pool;
```
[the remark this answers is not in the document]
Comment (abika5) [reply src/main/java/com/marklogic/contentpump/CompressedRDFReader.java:62]: Adding comments about removing the executor pool around the Jena parser calls.

File: src/main/java/com/marklogic/contentpump/RDFReader.java (hunk 3 above)
Comment (abika5) [reply src/main/java/com/marklogic/contentpump/RDFReader.java:1055]: Just the AsyncParser.asyncParse(in,lang,fsname,dest) should be able to handle all the types of RDFs. And this API call is currently being used for our in-memory RDF processing. The basic call asyncparse writes to streamRDF "dest" which is an event-based output. 

But our streamed RDF code flow is based on the parser output written and iterated via a stream. So to make minimal changes in the code flow I chose to use the stream-based API calls. 

We use 4 different API calls for the streaming RDF processing. For Ntriples and Nquads, the different API calls were suggested by the [Jena documentation](https://jena.apache.org/documentation/io/rdf-input.html#iterating-over-parser-output). For the Trig and else part, they are the same API calls but the difference is what they emit after parsing. Trig file outputs are emitted as quads and everything else like ttl, json, RDFXML file outputs are emitted as triples. Hence same API calls with different stream calls ( streamQuads()/streamTriples() )

File: pom.xml (hunk 7 above)
[the remark this answers is not in the document]
Comment (yunzvanessa) [reply pom.xml:801]: woodstox-core and stax-api both

**Review by yunzvanessa — CHANGES_REQUESTED** [review]

File: src/main/java/com/marklogic/contentpump/RDFReader.java [hunk 10]
```diff
@@ -348,6 +348,9 @@ protected void parse(String fsname, FSDataInputStream in)
 
     protected void loadModel(final String fsname, final InputStream in) throws IOException {
         if (dataset == null) {
+            // Previously during parsing, we create a runnable thread for each file and run them everytime.
```
[the remark this answers is not in the document]
Comment (abika5) [reply src/main/java/com/marklogic/contentpump/RDFReader.java:351]: In the compressedRDFreader, there is task creation for every zip entry. But in the RDFreader. We actually create threads directly and process each file. So changed the comment for compressedRDFreader

new Thread(jenaStreamingParser).start();

File: src/main/java/com/marklogic/contentpump/CompressedRDFReader.java [hunk 11]
```diff
@@ -124,6 +124,10 @@ protected void parse(String fsname, final InputStream in)
         throws IOException {
         if (dataset == null) {
             jenaStreamingParser = new RunnableParser(origFn, fsname, in, lang);
+            // Previously during parsing, we create a runnable thread for each file
```
[the remark this answers is not in the document]
Comment (abika5) [reply src/main/java/com/marklogic/contentpump/CompressedRDFReader.java:127]: You are right. I will make the change

**Review by yunzvanessa — APPROVED** [review]
Looks good. Thank you!

**Pull request merged**